### PR TITLE
Dikla Rikovitch RideShare Octos

### PR DESCRIPTION
**Pull request by diklaaharoni**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to choose how to calculate the in-progress trips I chose to write a helper method  that reject the in- progress trips and instead of looping on all the trips array I'm looping only on the finished trips array. 
Describe any examples of composition that you encountered in this project, if any |
Describe a _nominal test_ that you wrote for this assignment. | I wrote a test that assign the first driver whose status is available to the trip and change the driver's status to unavailable 
Describe an _edge case test_ that you wrote for this assignment. | I wrote a test that will raise an error if we try to finish trip that already finished
Describe a concept that you gained more clarity on as you worked on this assignment. | How to write a code with multiple classes

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -17,31 +18,54 @@ def initialize(input)
       @name = input[:name]
       @vehicle_id = input[:vin]
       @status = input[:status] == nil ? :AVAILABLE : input[:status]
-
       @trips = input[:trips] == nil ? [] : input[:trips]
     end
 
+    def finished_trips
+      trips.reject {|trip| trip.end_time == nil}
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:26]: I love this helper method! Good job breaking the problem up into multiple smaller pieces.

File: lib/traveler.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+module RideShare
+  class Traveler
+    attr_reader :trips
```
Comment (droberts-sea) [inline lib/traveler.rb:3]: I'm not sure what this class is for - it doesn't look like you use it anywhere.

It looks like maybe you were planning on making it a common superclass for `Driver` and `Passenger`, so that the `finished_trips` method doesn't need to be repeated. If so, this is an interesting idea, and I applaud the effort to keep your code DRY.

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -77,4 +77,99 @@
...
+        id: 7,
+        vin: 'a' * 17,
+        name: 'test driver',
+        trips: trips
+      }
+
+      driver = RideShare::Driver.new(driver_data)
+      driver.finished_trips.length.must_equal 2
+      trips.last.finish_trip!
+      driver.finished_trips.length.must_equal 3
```
Comment (droberts-sea) [inline specs/driver_spec.rb:119]: Good work testing this helper method - I like that this covers both with and without incomplete trips. I would also be interested in what happens when there are no complete trips.

File: specs/passenger_spec.rb [hunk 4]
```diff
@@ -75,4 +76,69 @@
...
+      trip = RideShare::Trip.new(trip_data)
+      @passenger = RideShare::Passenger.new(id: 9, name: "Merl Glover III", phone: "1-602-620-2330 x3723")
+      @passenger.add_trip(trip)
+    end
+
+    it "return the total amount of time that passenger has spent on their trips" do
+      @passenger.total_time.must_equal 1800
+    end
+
+    describe "finished_trips" do
```
Comment (droberts-sea) [inline specs/passenger_spec.rb:122]: This `describe` block is inside the `describe` block for `total_time` - not sure if this is the organization you intended.

File: specs/passenger_spec.rb [hunk 5]
```diff
@@ -75,4 +76,69 @@
...
+        rating: 3
+      }
+      trip = RideShare::Trip.new(trip_data)
+      @passenger = RideShare::Passenger.new(id: 9, name: "Merl Glover III", phone: "1-602-620-2330 x3723")
+      @passenger.add_trip(trip)
+    end
+
+    it "return the total amount of time that passenger has spent on their trips" do
+      @passenger.total_time.must_equal 1800
+    end
```
Comment (droberts-sea) [inline specs/passenger_spec.rb:120]: What if the passenger has no trips, or has an incomplete trip?

File: specs/trip_dispatch_spec.rb [hunk 6]
```diff
@@ -89,4 +89,90 @@
...
+        start_time: Time.now,
+        end_time: nil,
+        cost: nil,
+        rating: nil
+      }
+      @trip = RideShare::Trip.new(@trip_data)
+    end
+
+    it "Create a new instance of Trip" do
+      @trip.must_be_kind_of RideShare::Trip
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:109]: This test doesn't run any of `TripDispatcher`'s methods. Is it in the wrong place?

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -89,4 +89,90 @@
...
+
+      first_driver_avalable = dispatcher.request_trip(1).driver
+      first_driver_avalable.name.must_equal "Emory Rosenbaum"
+      first_driver_avalable.id.must_equal 2
+      first_driver_avalable.status.must_equal :UNAVAILABLE
+
+    end
+
+    it "Update the end time, cost and rating to be nil" do
+      dispatcher = RideShare::TripDispatcher.new
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:128]: Since `dispatcher` appears in multiple places, it would be a good candidate for an instance variable in a `before` block. It will be recreated from scratch before every test, so you don't need to worry about keeping track of how many trips other tests request.

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -89,4 +89,90 @@
...
+      driver.status.must_equal :UNAVAILABLE
+      passenger.name.must_equal "Nina Hintz Sr."
+      passenger.id.must_equal 1
+      trip.id.must_equal 601
+      trip.start_time.must_be_kind_of Time
+      trip.end_time.must_be_nil
+      trip.cost.must_be_nil
+      trip.rating.must_be_nil
+    end
+
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:174]: What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  dispatcher = RideShare::TripDispatcher.new
  dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

File: lib/driver.rb (hunk 1 above)
Comment (droberts-sea) [reply lib/driver.rb:26]: However, it seems like this should call `trip.finished?` instead of manually checking `trip.end_time == nil`. 

Bringing it back to the Metz reading, calling `trip.finished?` is an example of relying on _behavior_, whereas checking `trip.end_time` is relying on _data_.

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | no - In future projects, please work on having more frequent commits, with commit messages that describe how and why the code changed. If you're having trouble remembering to commit regularly, [you might find this article interesting](https://dev.to/rpalo/plan-your-commits)
Answer comprehension questions | mostly - By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | mostly - see inline
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes - good work
Complete tests | 
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** | Good job overall! This code is well-organized and easy to read. I'm especially impressed at the way you have broken your code up into small, manageable chunks. There are a couple places where your test coverage could be improved, but for where we're at in the course it's pretty solid.<br><br>I've left some nitpicks for you to review below, but in general I am quite happy with this submission. I'm excited to see what you come up with for Hotel. Keep up the hard work!
